### PR TITLE
Implemented a transitive closure that uses less intermediate containers

### PR DESCRIPTION
**Pull request by DavyLandman**
Solves #235

@msteindorfer maybe you can give pointers for a better way to use capsule? I especially miss an iterator over `<IValue, Set<IValue>>` or maybe `<IValue, Iterable<IValue>>`?

### CODE REVIEW

**Comment by jurgenvinju** [comment]
Thanks @DavyLandman! some questions: which algorithm is implemented here? Is it logically the same as previously? Do the current tests cover the possible changes that were made here? Is this safe to merge in that sense?

I think there are raw iterators for the mulltimap. I'll look that up. This makes the tuple2 optimization useless because most tuples will be stored in multimaps. That's cool! 

Still think we need to add an efficient merge/union for multimaps in capsule.

**Comment by DavyLandman** [comment]
I forgot how the algorithm is called, but it's roughly doing the same as before, except we use the Tries as an index (instead of rebuilding new indexes).

I've implemented this a few times in the past, and this was the strategy that always ended up being the fastest.

What happens is: we store the todo as an inverse index, as we'll need to index on the range on the original relation. so this way we reduce lookups on the original relation (only once per range) and just insert all the new pairs in the result set.

In essence, we do a compose, but only for the changed tuples, and since the left side of the compose needs to "index" on the range, we store the todo changed tuples in inverse.

**Comment by jurgenvinju** [comment]
> The thing I've forgotten is that if it's okay to compose to the original set, or that I need to compose to the new in-construction set.

I think it's ok. The algorithm is basically "depth-first-search" without anything specific to find but the nodes themselves. 

The inverse index is extremely important for average time complexity imho. Well spotted!

**Comment by DavyLandman** [comment]
if you want I'll try to summarize the algorithm in comments above it?

**Comment by jurgenvinju** [comment]
Am experimenting with `tupleIterator`, which does a native (fast) iteration over the multimap and calls a closure for every pair `a` and `b` in `{<a,b> | a <- keySet, b <- map[a]}` internally.

**Comment by DavyLandman** [comment]
@jurgenvinju looks good to me. I moved comment, and reduced the duplicate typebag in case the types are the same.

File: src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java
Comment (jurgenvinju) [inline src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java:1]: i will check check double check if we don't leak a mutable reference here ever.

Comment (jurgenvinju) [reply src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java:1]: Because the type bag is mutable in principle.

Comment (DavyLandman) [reply src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java:1]: I thought the TypeBag is immutable. as every `.increase` builds a new instance of `TypeBag`?

Comment (jurgenvinju) [reply src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java:1]: I read it wraps an immutable map that is replaced after every increase.

Comment (DavyLandman) [reply src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java:1]: the countMap is final and immutable
```
private final Map.Immutable<Type, Integer> countMap;
```

and every path of `increase` ends with:

```
return new TypeBag(newCountMap, ..);
```

**Comment by jurgenvinju** [comment]
Memory_telemetry

This is a large graph being transitively closed on a JVM -Xmx5000m running the Rascal console. A `rel[loc,loc]`, with the latest code from this PR. 

* It's using a lot less memory on short notice than before. The previous would get stuck after 3 minutes; this one after 30 minutes
* The bottleneck eventually seems GC, so I'm going to try with more memory this afternoon
* The internal iterator might still shave off memory overhead and GC time with it. There is a lot of trashing going on; but that must be the next "todo" list that is being made in every level of the recursion?

File: src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java
Comment (jurgenvinju) [reply src/main/java/io/usethesource/vallang/impl/persistent/PersistentHashIndexedBinaryRelation.java:1]: Yes thanks. I checked again. I saw another place yesterday where the increase was mutable, but it wasn't here. Better safe than sorry :-)

**Comment by jurgenvinju** [comment]
CPU load never goes above 70% btw.

**Comment by DavyLandman** [comment]
> The internal iterator might still shave off memory overhead and GC time with it. There is a lot of trashing going on; but that must be the next "todo" list that is being made in every level of the recursion?

yes, that's a pitty, the TODO gets thrashed all the time. I have never found a way around it. An alternative is keeping a list of nodes, and then just clearing that list. It does make it slower in that the `content.get` on the RHS is then called a lot more.

**Comment by jurgenvinju** [comment]
I think we might be missing out on depth-first approaches to this. The optimal algorithms in literature on sparse graphs (like ours) are not bread-first but depth-first, in paths. Our current implementation is breadth-first, in layers.  

Also we do not make use of the sparse nature of the hash-trie as much as we could, as an internal `compose` could also do. 
Leap-frog-trie join-like tricks could make closure a lot smaller and faster.

**Comment by jurgenvinju** [comment]
A depth-first search would also get rid of the large todo list, and trade it for recursion on the stack, and also avoid re-hashing nodes again and again that we already have visited. However, deep stacks are also harmful.... The fastest algorithms I know are fast because they switch strategies based on the size or shape of the incoming relation (large, small, componentized, layered, etc.).

**Comment by DavyLandman** [comment]
I've shared you a version of such an algorithm ;) check your PMs. It reduces the todo to a single collection that is never GCed (untill the end) and is max the size of the unique entries in the range of the relation.

**Pull request merged**